src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
                                 default(ImmutableArray<string>) :
                                 elementNames.ToImmutableAndFree();
 
-            NamedTypeSymbol tupleTypeOpt = null;
-            var elements = elementTypes.ToImmutableAndFree();
-            var locations = elementLocations.ToImmutableAndFree();
+            return new BoundTupleLiteral(node, boundArguments.ToImmutableAndFree(), elementNamesArray, 
+                                         declaredOrTargetTupleTypeOpt: null, hasErrors: hasErrors);
+        }
+
+        /// <summary>
+        /// Infer natural type for a tuple literal and store relevant information in the node.
+        /// 
+        /// A literal that has a natural type should not add any errors to <paramref name="diagnostics"/>
+        /// during the process.
+        /// </summary>
+        private BoundTupleLiteral AdjustTupleLiteralBasedOnNaturalTypeInference(BoundTupleLiteral tupleLiteral, DiagnosticBag diagnostics, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```

Comment:
@gafter Could you clarify please. Would you prefer a new field in a TypeInfo structure, if this are the only two options available?

---
In reply to: [62959049](https://github.com/dotnet/roslyn/pull/11201#discussion_r62959049) [](ancestors = 62959049)